### PR TITLE
go-client: add DialContext option to client

### PR DESCRIPTION
**Pull request by misterikkit**
### What does this PR achieve? Why do we need it?
This adds a configurable DialContext function to the client API, which
will be invoked to create HTTP/HTTPS connections.

Allowing a custom DialContext captures a wide variety of use cases and
behaviors by delegating the responsibility to the DialContext function.
Another benefit is that it is a minimal change in go-client itself.

For example, users can set dial timeouts, share connections, or
tunnel/proxy connections using custom logic.

### Does this PR fix issues?

Fixes #

### Notes for the reviewer

This is motivated by kubernetes network proxy use cases.

### CODE REVIEW

File: client/api/go-client/client.go [hunk 1]
```diff
@@ -189,11 +193,12 @@ func (c *Client) doBasic(req *http.Request) (*http.Response, error) {
 		<-c.throttle
 	}()
 
-	httpClient := &http.Client{}
-	if c.tlsClientConfig != nil {
-		httpClient.Transport = &http.Transport{
-			TLSClientConfig: c.tlsClientConfig,
-		}
+	httpClient := &http.Client{
+		CheckRedirect: c.checkRedirect,
```
Comment (phlogistonjohn) [inline client/api/go-client/client.go:197]: Looks like this is redundant with `httpClient.CheckRedirect = c.checkRedirect` below.

Comment (misterikkit) [reply client/api/go-client/client.go:197]: Oops, good catch.

**Comment by misterikkit** [comment]
Thanks for the quick review! :smile: 

> In addition, I love seeing the negative test case, but I'd like to a positive use too.

I added a positive test, and then realized the negative test needed to be a little more realistic (and the test server doesn't support `/hello`)

**Comment by misterikkit** [comment]
Someone pointed out to me that my change actually discards the default http transport timeouts that were implicitly used before. That is addressed now.

**Comment by phlogistonjohn** [comment]
Looks like the CI is still failing. This time it may be because you're using apis only in newer Go versions? We still test with 1.10. Let me know if you think you can make it work on 1.10 or not.

File: client/api/go-client/default_transport_clone_go112.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+//
+// You may not use this file except in compliance with those terms.
+//
+
+package client
+
+import "net/http"
+
+// make a copy of the default transport at package initialization time
+var defaultTransportCopy = *(http.DefaultTransport.(*http.Transport))
```
Comment (liggitt) [inline client/api/go-client/default_transport_clone_go112.go:20]: bah, this fails with:
```
-- Running: 030-govet.sh --
# github.com/heketi/heketi/client/api/go-client
client/api/go-client/default_transport_clone_go112.go:20: variable declaration copies lock value to defaultTransportCopy: net/http.Transport contains sync.Mutex
failed ./tests/030-govet.sh [2]
```

you'll need to do the field-by-field copy here as well

```go
var copyOfDefaultTransport *http.Transport

func init() {
  copyOfDefaultTransport = copyTransport(http.DefaultTransport.(*http.Transport))
}

func copyTransport(in *http.Transport) *http.Transport {
  return &http.Transport{
    Proxy: in.Proxy, 
    ...
  }
}

func defaultTransportClone() *http.Transport {
  return copyTransport(copyOfDefaultTransport)
}
```

Comment (misterikkit) [reply client/api/go-client/default_transport_clone_go112.go]: Good catch, and great presubmit coverage.

**Comment by misterikkit** [comment]
This time it seems like an infra problem failed the test, so let's try again.

/retest

**Comment by misterikkit** [comment]
I will do a rebase and try to get these tests passing today or early next week

**Comment by misterikkit** [comment]
Rebased. Hopefully presubmit works this time.

**Comment by misterikkit** [comment]
@phlogistonjohn This should be working on go1.10, but I think the CI is flaking? I've just rebased so hopefully that fixes it.

**Comment by misterikkit** [comment]
Looks like tests failed with this error, and I'm not sure if that's related to this PR.

```
16:30:17 ==> PASSED TestEnabledTLS
16:30:17 ==> Started Fri May 29 23:38:45 BST 2020
16:30:17 ==> Ended Sat May 30 00:30:17 BST 2020
16:30:17 ==> -- Passing Tests: TestSelfTest TestVolumeNotDeletedWhenNodeIsDown TestUpgrade TestEnabledTLS
16:30:17 ==> -- Failing Tests: TestSmokeTest TestErrorHandling
16:30:17 ==> FAILED
```

**Comment by obnoxxx** [comment]
retest this please

**Comment by humblec** [comment]
>This should be working on go1.10
@misterikkit maybe its also good to try on a later version of go binary.

**Comment by phlogistonjohn** [comment]
retest this please

**Comment by misterikkit** [comment]
Thanks for the comments.

> @misterikkit The PR looks good to me! maybe its also good to try on a later version of go binary. say >=1.13

Back on Apr 9 this was working on both go1.10 and go1.13. I have rebased the change on master and manually re-tested it on the following go versions,
- 1.10.8
- 1.11.13
- 1.12.17
- 1.13.11
- 1.14.4

I ran tests like so,
```
PATH="$(go1.10.8 env GOROOT)/bin:$PATH" make clean_vendor vendor test
```

Regarding the failing `centos-ci` presubmit, do I re-trigger it by commenting "/retest" on this PR?

**Comment by humblec** [comment]
retest this please

**Comment by humblec** [comment]
@phlogistonjohn it seems that the failure in CI is spurious or unrelated to this PR ? if there is no way to get rid of this spurious failure in the CI, can we get this merged manually with some priority?

**Comment by phlogistonjohn** [comment]
> @phlogistonjohn it seems that the failure in CI is spurious or unrelated to this PR ? if there is no way to get rid of this spurious failure in the CI, can we get this merged manually with some priority?

I have restarted it a few times myself. Unfortuantely, the CI is not typically this flakey which makes me wonder if we're missing something subtle about the changes that decrease the reliability of the tests. I will try to dig into it some more and see what I can find out.

**Comment by phlogistonjohn** [comment]
To expand on my previous comment: many of the failures I am seeing are failures to connect to the server. If the failures were totally unrelated to client/server comms I'd me more inclined to ignore it as flakiness, but the kind of failure has me concerned.

**Comment by misterikkit** [comment]
Good catch, John. It's possible I inadvertently changed the TCP timeouts in
this PR. Let me take a closer look.

>

**Comment by misterikkit** [comment]
@phlogistonjohn  @humblec  I can see what is different now.

Before this change,
* if `tlsClientConfig` is nil, then `http.DefaultTransport` is used.
* if `tlsClientConfig` is not nil, then a "zero value" `http.Transport` is used.

With this change, `http.DefaultTransport` is always used.

Interesting differences between default and zero

field | default | zero
-|-|-
Proxy | `ProxyFromEnvironment` | nil (no proxy)
DialContext | timeout 30s, keepalive 30s | timeout none, keepalive 15s
ForceAttemptHttp2 | true | false

I could change this PR to preserve the current logic, but I cannot tell if that logic is intentional. Either way, I'll upload a patch to see if it fixes the CI.

**Comment by phlogistonjohn** [comment]
It probably was not intentional, but could explain some of the CI failures, esp. wrt the SSL test suite! Thanks for looking into it with that level of detail.

**Comment by misterikkit** [comment]
This time the centos-ci failures seem to all be "no route to host" which is less likely to be related here.

```
12:23:22 [cmdexec] WARNING 2020/06/15 20:23:22 Failed to create SSH connection to 192.168.10.100:22: dial tcp 192.168.10.100:22: connect: no route to host
12:23:22 [cmdexec] ERROR 2020/06/15 20:23:22 heketi/executors/cmdexec/peer.go:80:cmdexec.(*CmdExecutor).GlusterdCheck: dial tcp 192.168.10.100:22: connect: no route to host
```

**Comment by humblec** [comment]
```
19:21:05 [cmdexec] ERROR 2020/06/15 20:21:05 heketi/pkg/remoteexec/log/commandlog.go:56:log.(*CommandLogger).Error: Failed to run command [systemctl status glusterd] on [192.168.10.100:22]: Err[Process exited with status 4]: Stdout []: Stderr [Unit glusterd.service could not be found.
19:21:05 ]
..

19:21:56 [heketi] ERROR 2020/06/15 20:21:56 heketi/apps/glusterfs/operations_manage.go:257:glusterfs.RunOperation: Remove Device Build Failed: The target exists, contains other items, or is in use.
..
19:22:11 ==> FAILED TestVolumeNotDeletedWhenNodeIsDown
...

9:22:35 ERROR
19:22:35 
19:22:35 ======================================================================
19:22:35 ERROR: test_create_cluster_tls_cert (__main__.TestTLS)
19:22:35 ----------------------------------------------------------------------
19:22:35 Traceback (most recent call last):
19:22:35   File "test_tls.py", line 89, in setUp
19:22:35     self.heketi = HeketiServer().start()
19:22:35   File "test_tls.py", line 59, in start
19:22:35     raise SetupError('Heketi server failed to start')
19:22:35 SetupError: Heketi server failed to start
19:22:35 
...
19:22:37 --- FAILED: test_tls.py client_tls_test
19:22:37 ==> FAILED TestEnabledTLS
19:22:37 ==> TEARDOWN TestEnabledTLS
19:22:37 ./run.sh: line 154: teardown.sh: command not found
19:22:37 ==> Started Mon Jun 15 19:03:31 BST 2020
19:22:37 ==> Ended Mon Jun 15 20:22:37 BST 2020
19:22:37 ==> -- Passing Tests: TestSelfTest TestUpgrade
19:22:37 ==> -- Failing Tests: TestSmokeTest TestErrorHandling TestVolumeNotDeletedWhenNodeIsDown TestEnabledTLS
19:22:37 ==> FAILED
19:22:37 make[1]: *** [test] Error 1
19:22:37 make[1]: Leaving directory `/root/go/src/github.com/heketi/heketi/tests/functional'
19:22:37 make: *** [test-functional] Error 2
19:23:04 [heketi] INFO 2020/06/15 20:23:04 Starting Node Health Status refresh
19:23:04 [cmdexec] INFO 2020/06/15 20:23:04 Check Glusterd service status in node 192.168.10.100
19:23:22 [cmdexec] WARNING 2020/06/15 20:23:22 Failed to create SSH connection to 192.168.10.100:22: dial tcp 192.168.10.100:22: connect: no route to host
...
````

Above are some of the errors I see in the CI :

@phlogistonjohn

**Comment by liggitt** [comment]
@misterikkit any update here?

**Comment by humblec** [comment]
@heketi/maintainers can we get some traction or help here ? this is an important PR to get in. Please prioritize this.

**Comment by humblec** [comment]
@phlogistonjohn @raghavendra-talur @obnoxxx can we help in full priority to get this closed as this is an important PR as mentioned earlier ?

**Comment by obnoxxx** [comment]
If you look at the description, this PR looks very innocent and not very urgent.
But it is blocking a change in kubernetes (https://github.com/kubernetes/kubernetes/pull/91785) that is needed to fix a CVE issue (https://github.com/kubernetes/kubernetes/pull/91785).

Hence the increased sense of urgency.

**Comment by raghavendra-talur** [comment]
@centos-ci retest this please

**Comment by raghavendra-talur** [comment]
@misterikkit I did a small experiment to see if this PR is causing the `http: Accept error: accept tcp [::]:8080: accept4: too many open files;` errors and looks like it is.

In the PR #1768 , I initially had two small commits that making reading through test logs easier and reduce the number of tests that are run. The tests passed.

Then, I cherry-picked the two commits from this PR and ran the tests again. The tests failed with the following error:

```
23:06:46 === RUN   TestAdminModes
23:08:17 --- FAIL: TestAdminModes (90.81s)
23:08:17     assert.go:44: 
23:08:17         
ASSERT:	func (github.com/heketi/heketi/pkg/testutils.(*ClusterEnv).SetupWithCluster) 0x713527
23:08:17         
	File /root/go/src/github.com/heketi/heketi/pkg/testutils/cluster.go:152
23:08:17         
	Info: [expected err == nil, got: Get http://localhost:8080/queue/22e7995bf5bc8d744a339b19a89a3bb3: dial tcp: lookup localhost: device or resource busy]
```

I am guessing that the default transport clone resets/set a timeout which is leaving the fds open longer than earlier. Please investigate if there is a simpler way to make the changes in the PR.

**Comment by phlogistonjohn** [comment]
As a point I'd like anyone interested in this discussion to take a look at #1771. I came up with this alternate approach late last week. There's less change in heketi directly but should allow for lots more customization by consumers of the heketi client libs. Please feel free to comment on the general approach here or on #1771 and specifics directly on that PR.

This could be an alternative to this PR or be used to complement it.

**Comment by obnoxxx** [comment]
@misterikkit - any comments?
Is it still interesting to get this into heketi?
@liggitt , @gnufied

**Comment by misterikkit** [comment]
Yes, still interested in re-working this or merging https://github.com/heketi/heketi/pull/1771 in its place. Unfortunately I haven't had time to focus on it in the last month.

**Comment by obnoxxx** [comment]
> Yes, still interested in re-working this or merging #1771 in its place. Unfortunately I haven't had time to focus on it in the last month.

@misterikkit ok, cool. Therefore, it might be good to give #1771 a closer look. If that serves your needs (I think it should), then we can indeed independently bring that forward, and thereby enable fixing the issue in kubernetes, giving you a bit more time.

**Comment by misterikkit** [comment]
Closing this because  #1771  solves it.

**Pull request closed without merging**